README.md
# <img alt="SageMaker" src="branding/icon/sagemaker-banner.png" height="100">

# SageMaker TensorFlow Serving Container

SageMaker TensorFlow Serving Container is an a open source project that builds
docker images for running TensorFlow Serving on
[Amazon SageMaker](https://aws.amazon.com/documentation/sagemaker/).

Supported versions of TensorFlow: ``1.4.1``, ``1.5.0``, ``1.6.0``, ``1.7.0``, ``1.8.0``, ``1.9.0``, ``1.10.0``, ``1.11.0``, ``1.12.0``, ``1.13.1``, ``1.14.0``, ``1.15.0``, ``2.0.0``.

Supported versions of TensorFlow for Elastic Inference: ``1.11.0``, ``1.12.0``, ``1.13.1``, ``1.14.0``.

ECR repositories for SageMaker built TensorFlow Serving Container:

- `'tensorflow-inference'` for any new version starting with ``1.13.0`` in the following AWS accounts:
  - `"871362719292"` in `"ap-east-1"`;
  - `"217643126080"` in `"me-south-1"`;
  - `"886529160074"` in `"us-iso-east-1"`;
  - `"763104351884"` in other SageMaker public regions.
- `'sagemaker-tensorflow-serving'` for ``1.4.1``, ``1.5.0``, ``1.6.0``, ``1.7.0``, ``1.8.0``, ``1.9.0``, ``1.10.0``, ``1.11.0``, ``1.12.0`` versions in the following AWS accounts:
  - `"057415533634"` in `"ap-east-1"`;
  - `"724002660598"` in `"me-south-1"`;
  - `"520713654638"` in other SageMaker public regions.

ECR repositories for SageMaker built TensorFlow Serving Container for Elastic Inference:

- `'tensorflow-inference-eia'` for any new version starting with ``1.14.0`` in the same AWS accounts as TensorFlow Serving Container for newer TensorFlow versions listed above;
- `'sagemaker-tensorflow-serving-eia'` for ``1.11.0``, ``1.12.0``, ``1.13.1`` versions in the same AWS accounts as TensorFlow Serving Container for older TensorFlow versions listed above.

This documentation covers building and testing these docker images.

For information about using TensorFlow Serving on SageMaker, see:
[Deploying to TensorFlow Serving Endpoints](https://sagemaker.readthedocs.io/en/stable/frameworks/tensorflow/deploying_tensorflow_serving.html)
in the [SageMaker Python SDK](https://github.com/aws/sagemaker-python-sdk) documentation.

For notebook examples, see: [Amazon SageMaker Examples](https://github.com/awslabs/amazon-sagemaker-examples).

## Table of Contents

1. [Getting Started](#getting-started)
2. [Building your image](#building-your-image)
3. [Running the tests](#running-the-tests)
4. [Pre/Post-Processing](#pre/post-processing)
5. [Deploying a TensorFlow Serving Model](#deploying-a-tensorflow-serving-model)
6. [Deploying to Multi-Model Endpoint](#deploying-to-multi-model-endpoint)

## Getting Started

### Prerequisites

Make sure you have installed all of the following prerequisites on your
development machine:

- [Docker](https://www.docker.com/)
- [AWS CLI](https://aws.amazon.com/cli/)

For testing, you will also need:

- [Python 3.6](https://www.python.org/)
- [tox](https://tox.readthedocs.io/en/latest/)
- [npm](https://npmjs.org/)
- [jshint](https://jshint.com/about/)

To test GPU images locally, you will also need:

- [nvidia-docker](https://github.com/NVIDIA/nvidia-docker)

**Note:** Some of the build and tests scripts interact with resources in your AWS account. Be sure to
set your default AWS credentials and region using `aws configure` before using these scripts.

## Building your image

Amazon SageMaker uses Docker containers to run all training jobs and inference endpoints.

The Docker images are built from the Dockerfiles in
[docker/](https://github.com/aws/sagemaker-tensorflow-serving-container/tree/master/docker).

The Dockerfiles are grouped based on the version of TensorFlow Serving they support. Each supported
processor type (e.g. "cpu", "gpu", "ei") has a different Dockerfile in each group.

To build an image, run the `./scripts/build.sh` script:

```bash
./scripts/build.sh --version 1.13 --arch cpu
./scripts/build.sh --version 1.13 --arch gpu
./scripts/build.sh --version 1.13 --arch eia
```


If your are testing locally, building the image is enough. But if you want to your updated image
in SageMaker, you need to publish it to an ECR repository in your account. The
`./scripts/publish.sh` script makes that easy:

```bash
./scripts/publish.sh --version 1.13 --arch cpu
./scripts/publish.sh --version 1.13 --arch gpu
./scripts/publish.sh --version 1.13 --arch eia
```

Note: this will publish to ECR in your default region. Use the `--region` argument to
specify a different region.

### Running your image in local docker

You can also run your container locally in Docker to test different models and input
inference requests by hand. Standard `docker run` commands (or `nvidia-docker run` for
GPU images) will work for this, or you can use the provided `start.sh`
and `stop.sh` scripts:

```bash
./scripts/start.sh [--version x.xx] [--arch cpu|gpu|eia|...]
./scripts/stop.sh [--version x.xx] [--arch cpu|gpu|eia|...]
```

When the container is running, you can send test requests to it using any HTTP client. Here's
and an example using the `curl` command:

```bash
curl -X POST --data-binary @test/resources/inputs/test.json \
     -H 'Content-Type: application/json' \
     -H 'X-Amzn-SageMaker-Custom-Attributes: tfs-model-name=half_plus_three' \
     http://localhost:8080/invocations
```

Additional `curl` examples can be found in `./scripts/curl.sh`.

## Running the tests

The package includes automated tests and code checks. The tests use Docker to run the container
image locally, and do not access resources in AWS. You can run the tests and static code
checkers using `tox`:

```bash
tox
```

To run local tests against a single container or with other options, you can use the following command:

```bash
python -m pytest test/integration/local
    [--docker-name-base <docker_name_base>]
    [--framework-version <framework_version>]
    [--processor-type <processor_type>]
```

To test against Elastic Inference with Accelerator, you will need an AWS account, publish your built image to ECR repository and run the following command:

    tox -e py36 -- test/integration/sagemaker/test_ei.py
        [--repo <ECR_repository_name>]
        [--instance-types <instance_type>,...]
        [--accelerator-type <accelerator_type>]
        [--versions <version>,...]

For example:

    tox -e py36 -- test/integration/sagemaker/test_ei.py \
        --repo sagemaker-tensorflow-serving-eia \
        --instance_type ml.m5.xlarge \
        --accelerator-type ml.eia1.medium \
        --versions 1.13.0


## Pre/Post-Processing

**NOTE: There is currently no support for pre-/post-processing with multi-model containers.**

SageMaker TensorFlow Serving Container supports the following Content-Types for requests:

* `application/json` (default)
* `text/csv`
* `application/jsonlines`

And the following content types for responses:

* `application/json` (default)
* `application/jsonlines`

The container will convert data in these formats to [TensorFlow Serving REST API](https://www.tensorflow.org/tfx/serving/api_rest) requests,
and will send these requests to the default serving signature of your SavedModel bundle.

You can also add customized Python code to process your input and output data. To use this feature, you need to:
1. Add a python file named `inference.py` to the code directory inside your model archive.
2. In `inference.py`, implement either a pair of `input_handler` and `output_handler` functions or a single `handler` function. Note that if `handler` function is implemented, `input_handler` and `output_handler` will be ignored.

To implement pre/post-processing handler(s), you will need to make use of the `Context` object created by Python service. The `Context` is a `namedtuple` with following attributes:
- `model_name (string)`: the name of the model you will to use for inference, for example 'half_plus_three'
- `model_version (string)`: version of the model, for example '5'
- `method (string)`: inference method, for example, 'predict', 'classify' or 'regress', for more information on methods, please see [Classify and Regress API](https://www.tensorflow.org/tfx/serving/api_rest#classify_and_regress_api) and [Predict API](https://www.tensorflow.org/tfx/serving/api_rest#predict_api)
- `rest_uri (string)`: the TFS REST uri generated by the Python service, for example, 'http://localhost:8501/v1/models/half_plus_three:predict'
- `grpc_port (string)`: the GRPC port number generated by the Python service, for example, '9000'
- `custom_attributes (string)`: content of 'X-Amzn-SageMaker-Custom-Attributes' header from the original request, for example, 'tfs-model-name=half_plus_three,tfs-method=predict'
- `request_content_type (string)`: the original request content type, defaulted to 'application/json' if not provided
- `accept_header (string)`: the original request accept type, defaulted to 'application/json' if not provided
- `content_length (int)`: content length of the original request

Here's a code example implementing `input_handler` and `output_handler`. By providing these, the Python service will post the request to TFS REST uri with the data pre-processed by `input_handler` and pass the response to `output_handler` for post-processing.

```python
import json

def input_handler(data, context):
    """ Pre-process request input before it is sent to TensorFlow Serving REST API
    Args:
        data (obj): the request data, in format of dict or string
        context (Context): an object containing request and configuration details
    Returns:
        (dict): a JSON-serializable dict that contains request body and headers
    """
    if context.request_content_type == 'application/json':
        # pass through json (assumes it's correctly formed)
        d = data.read().decode('utf-8')
        return d if len(d) else ''

    if context.request_content_type == 'text/csv':
        # very simple csv handler
        return json.dumps({
            'instances': [float(x) for x in data.read().decode('utf-8').split(',')]
        })

    raise ValueError('{{"error": "unsupported content type {}"}}'.format(
        context.request_content_type or "unknown"))


def output_handler(data, context):
    """Post-process TensorFlow Serving output before it is returned to the client.
    Args:
        data (obj): the TensorFlow serving response
        context (Context): an object containing request and configuration details
    Returns:
        (bytes, string): data to return to client, response content type
    """
    if data.status_code != 200:
        raise ValueError(data.content.decode('utf-8'))

    response_content_type = context.accept_header
    prediction = data.content
    return prediction, response_content_type
```

Here's another code example implementing `input_handler` and `output_handler` to format image data into a TFS request that expects image data as an encoded string rather than as a numeric tensor:

```python
import base64
import io
import json
import requests

def input_handler(data, context):
    """ Pre-process request input before it is sent to TensorFlow Serving REST API

    Args:
        data (obj): the request data stream
        context (Context): an object containing request and configuration details

    Returns:
        (dict): a JSON-serializable dict that contains request body and headers
    """

    if context.request_content_type == 'application/x-image':
        payload = data.read()
        encoded_image = base64.b64encode(payload).decode('utf-8')
        instance = [{"b64": encoded_image}]
        return json.dumps({"instances": instance})
    else:
        _return_error(415, 'Unsupported content type "{}"'.format(
            context.request_content_type or 'Unknown'))


def output_handler(response, context):
    """Post-process TensorFlow Serving output before it is returned to the client.

    Args:
        response (obj): the TensorFlow serving response
        context (Context): an object containing request and configuration details

    Returns:
        (bytes, string): data to return to client, response content type
    """
    if response.status_code != 200:
        _return_error(response.status_code, response.content.decode('utf-8'))
    response_content_type = context.accept_header
    prediction = response.content
    return prediction, response_content_type


def _return_error(code, message):
    raise ValueError('Error: {}, {}'.format(str(code), message))
```

The `input_handler` above creates requests that match the input of the following TensorFlow Serving SignatureDef, displayed
using the TensorFlow `saved_model_cli`:

```
signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['image_bytes'] tensor_info:
        dtype: DT_STRING
        shape: (-1)
        name: input_tensor:0
  The given SavedModel SignatureDef contains the following output(s):
    outputs['classes'] tensor_info:
        dtype: DT_INT64
        shape: (-1)
        name: ArgMax:0
    outputs['probabilities'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 1001)
        name: softmax_tensor:0
  Method name is: tensorflow/serving/predict
```


There are occasions when you might want to have complete control over the request handler. For example, making TFS request (REST or GRPC) to one model, and then making a request to a second model. In this case, you may implement the `handler` instead of the `input_handler` and `output_handler` pair:

```python
import json
import requests


def handler(data, context):
    """Handle request.
    Args:
        data (obj): the request data
        context (Context): an object containing request and configuration details
    Returns:
        (bytes, string): data to return to client, (optional) response content type
    """
    processed_input = _process_input(data, context)
    response = requests.post(context.rest_uri, data=processed_input)
    return _process_output(response, context)


def _process_input(data, context):
    if context.request_content_type == 'application/json':
        # pass through json (assumes it's correctly formed)
        d = data.read().decode('utf-8')
        return d if len(d) else ''

    if context.request_content_type == 'text/csv':
        # very simple csv handler
        return json.dumps({
            'instances': [float(x) for x in data.read().decode('utf-8').split(',')]
        })

    raise ValueError('{{"error": "unsupported content type {}"}}'.format(
        context.request_content_type or "unknown"))


def _process_output(data, context):
    if data.status_code != 200:
        raise ValueError(data.content.decode('utf-8'))

    response_content_type = context.accept_header
    prediction = data.content
    return prediction, response_content_type
```

You can also bring in external dependencies to help with your data processing. There are 2 ways to do this:
1. If your model archive contains `code/requirements.txt`, the container will install the Python dependencies at runtime using `pip install -r`.
2. If you are working in a network-isolation situation or if you don't want to install dependencies at runtime everytime your Endpoint starts or Batch Transform job runs, you may want to put pre-downloaded dependencies under `code/lib` directory in your model archive, the container will then add the modules to the Python path. Note that if both `code/lib` and `code/requirements.txt` are present in the model archive, the `requirements.txt` will be ignored.

Your untarred model directory structure may look like this if you are using `requirements.txt`:

        model1
            |--[model_version_number]
                |--variables
                |--saved_model.pb
        model2
            |--[model_version_number]
                |--assets
                |--variables
                |--saved_model.pb
        code
            |--inference.py
            |--requirements.txt

Your untarred model directory structure may look like this if you have downloaded modules under `code/lib`:

        model1
            |--[model_version_number]
                |--variables
                |--saved_model.pb
        model2
            |--[model_version_number]
                |--assets
                |--variables
                |--saved_model.pb
        code
            |--lib
                |--external_module
            |--inference.py

## Deploying a TensorFlow Serving Model

To use your TensorFlow Serving model on SageMaker, you first need to create a SageMaker Model. After creating a SageMaker Model, you can use it to create [SageMaker Batch Transform Jobs](https://docs.aws.amazon.com/sagemaker/latest/dg/how-it-works-batch.html)
 for offline inference, or create [SageMaker Endpoints](https://docs.aws.amazon.com/sagemaker/latest/dg/how-it-works-hosting.html) for real-time inference.


### Creating a SageMaker Model

A SageMaker Model contains references to a `model.tar.gz` file in S3 containing serialized model data, and a Docker image used to serve predictions with that model.

You must package the contents in a model directory (including models, inference.py and external modules) in .tar.gz format in a file named "model.tar.gz" and upload it to S3. If you're on a Unix-based operating system, you can create a "model.tar.gz" using the `tar` utility:

```
tar -czvf model.tar.gz 12345 code
```

where "12345" is your TensorFlow serving model version which contains your SavedModel.

After uploading your `model.tar.gz` to an S3 URI, such as `s3://your-bucket/your-models/model.tar.gz`, create a [SageMaker Model](https://docs.aws.amazon.com/sagemaker/latest/dg/API_CreateModel.html) which will be used to generate inferences. Set `PrimaryContainer.ModelDataUrl` to the S3 URI where you uploaded the `model.tar.gz`, and set `PrimaryContainer.Image` to an image following this format:

```
520713654638.dkr.ecr.{REGION}.amazonaws.com/sagemaker-tensorflow-serving:{SAGEMAKER_TENSORFLOW_SERVING_VERSION}-{cpu|gpu}
```

```

CHANGELOG.md
# Changelog

## v1.8.4 (2021-06-30)

### Bug Fixes and Other Changes

 * modify the way port number passing

## v1.8.3 (2021-04-26)

### Bug Fixes and Other Changes

 * Create test_multi_tfs.py

## v1.8.2 (2021-04-13)

### Bug Fixes and Other Changes

 * Update test_container.py

## v1.8.1 (2021-04-07)

### Bug Fixes and Other Changes

 * set OMP_NUM_THREADS default value to 1
 * Wait tfs before starting gunicorn

## v1.8.0 (2021-03-23)

### Features

 * expose tunable parameters to support multiple tfs
 * universal requirements.txt and inference.py

### Bug Fixes and Other Changes

 * To fix python sdk repo localmode tfs tests failing issue
 * Fix TFS Deploy codebuild issue
 * Remove the 1s lock for inference
 * install boto3
 * extending the serve.py with GUNICORN workers and threads
 * adding isnumeric filter to find_model_versions function.
 * return information of all models
 * Support multiple Accept types

### Documentation Changes

 * fix typo in README.md
 * updating readme with the right context object
 * fix broken link in README

## v1.7.0 (2020-07-29)

### Features

 * add model_version_policy to model config

## v1.6.25 (2020-07-23)

### Bug Fixes and Other Changes

 * change single quotes to double quotes

## v1.6.24 (2020-07-17)

### Bug Fixes and Other Changes

 * update MME Pre/Post-Processing model and script paths
 * increasing max_retry for model availability check
 * multi-model-endpoint support

### Documentation Changes

 * update README for multi-model endpoint

## v1.6.23.post0 (2020-06-25)

### Testing and Release Infrastructure

 * add issue templates

## v1.6.23 (2020-06-11)

### Bug Fixes and Other Changes

 * error code files for TFS2.2 (copied from pip install tf)

## v1.6.22.post0 (2020-05-13)

### Testing and Release Infrastructure

 * Test against py37 in buildspec

## v1.6.22 (2020-04-20)

### Bug Fixes and Other Changes

 * TF EIA 1.15 and 2.0

## v1.6.21 (2020-04-16)

### Bug Fixes and Other Changes

 * Replaced deprecated aws ecr get-login by get-login-password

## v1.6.20 (2020-04-03)

### Bug Fixes and Other Changes

 * Updating Pyyaml and Awscli version for tf1.15
 * upgrade Pillow version in test example

## v1.6.19 (2020-04-01)

### Bug Fixes and Other Changes

 * Allowing arguments for deep_learning_container.py for tf

## v1.6.18 (2020-03-26)

### Bug Fixes and Other Changes

 * Adding of deep_learning_container.py

## v1.6.17 (2020-02-18)

### Bug Fixes and Other Changes

 * update: Remove multi-model label from CPU containers

## v1.6.16 (2020-02-17)

### Bug Fixes and Other Changes

 * update r2.0.1 dockerfiles
 * add 2.1 dockerfiles and tensorflow artifacts
 * update for 1.15.2

## v1.6.15 (2020-02-04)

### Bug Fixes and Other Changes

 * return on_delete method when model successfully deleted
 * validate tensorflow model version number in model path

## v1.6.14.post0 (2020-01-20)

### Documentation changes

 * document that pre-/post-processing is not supported with multi-model

## v1.6.14 (2020-01-10)

### Bug fixes and other changes

 * Add __init__.py to fix unit test
 * update: Update buildspec for TF 1.15 and 2.0

## v1.6.13 (2020-01-08)

### Bug fixes and other changes

 * update copyright year in license header

## v1.6.12.post0 (2020-01-03)

### Documentation changes

 * update Readme with correct TF versions.

## v1.6.12 (2020-01-02)

### Bug fixes and other changes

 * update: Release TF 1.15 and TF 2.0 dockerfiles

## v1.6.11 (2019-12-17)

### Bug fixes and other changes

 * check container is ready in tests

## v1.6.10 (2019-12-13)

### Bug fixes and other changes

 * increase attempts to allow for large gpu images
 * ping check and model status check

## v1.6.9 (2019-11-25)

### Bug fixes and other changes

 * Update EI Dockerfile 1.14 with New Health check, and new binaries

## v1.6.8 (2019-10-25)

### Bug fixes and other changes

 * update publish-all.sh to match the versions in build-all.sh
 * upgrade pillow to 6.2.0 in requirements.txt test
 * use regional endpoint for STS in builds and tests
 * merge dockerfiles

## v1.6.7 (2019-10-22)

### Bug fixes and other changes

 * update instance type region availability

## v1.6.6 (2019-10-16)

### Bug fixes and other changes

 * mme improvements

## v1.6.5 (2019-09-30)

### Bug fixes and other changes

 * "change: merge asimov branch to master branch (#80)"
 * merge asimov branch to master branch

## v1.6.4 (2019-09-17)

### Bug fixes and other changes

 * TFS_SHORT_VERSION explicitly defined in dockerfile

## v1.6.3 (2019-09-13)

### Bug fixes and other changes

 * add Dockerfile for 1.14 in py3 with EIA

## v1.6.2 (2019-09-09)

### Bug fixes and other changes

 * require defaul model env var
 * default SAGEMAKER_TFS_DEFAULT_MODEL_NAME to string value
 * remove unused fixture

## v1.6.1 (2019-08-15)

### Bug fixes and other changes

 * Update no-p2 and no-p3 regions

## v1.6.0 (2019-08-09)

### Features

 * delete models
 * different  invocation url
 * list models
 * add model

### Bug fixes and other changes

 * update request/response model and url key names
 * README changes

### Documentation changes

 * add batching to readme

## v1.5.1 (2019-06-21)

### Bug fixes and other changes

 * use nvidia-docker for local gpu tests

## v1.5.0 (2019-06-20)

### Features

 * add 1.13 EIA containers

## v1.4.6 (2019-06-18)

### Bug fixes and other changes

 * fix broken ei tests

## v1.4.5 (2019-06-17)

### Bug fixes and other changes

 * add batch transform integration test

## v1.4.4 (2019-06-12)

### Bug fixes and other changes

 * move SageMaker tests to release build

## v1.4.3 (2019-06-12)

### Bug fixes and other changes

 * use p2.xlarge by default in tests

## v1.4.2 (2019-06-12)

### Bug fixes and other changes

 * add Tensorflow 1.13

## v1.4.1 (2019-06-11)

### Bug fixes and other changes

 * make tox run any pytest tests

## v1.4.0 (2019-06-03)

### Features

 * support jsonlines output

### Documentation changes

 * update README.md for EI image


docker/2.1/sagemaker/serve.py
# Copyright 2018-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://aws.amazon.com/apache2.0/
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

import logging
import os
import re
import signal
import subprocess
import tfs_utils

from contextlib import contextmanager

logging.basicConfig(level=logging.INFO)
log = logging.getLogger(__name__)

JS_PING = "js_content ping"
JS_INVOCATIONS = "js_content invocations"
GUNICORN_PING = "proxy_pass http://gunicorn_upstream/ping"
GUNICORN_INVOCATIONS = "proxy_pass http://gunicorn_upstream/invocations"

PYTHON_LIB_PATH = "/opt/ml/model/code/lib"
REQUIREMENTS_PATH = "/opt/ml/model/code/requirements.txt"
INFERENCE_PATH = "/opt/ml/model/code/inference.py"


class ServiceManager(object):
    def __init__(self):
        self._state = "initializing"
        self._nginx = None
        self._tfs = None
        self._gunicorn = None
        self._gunicorn_command = None
        self._enable_python_service = os.path.exists(INFERENCE_PATH)
        self._tfs_version = os.environ.get("SAGEMAKER_TFS_VERSION", "1.13")
        self._nginx_http_port = os.environ.get("SAGEMAKER_BIND_TO_PORT", "8080")
        self._nginx_loglevel = os.environ.get("SAGEMAKER_TFS_NGINX_LOGLEVEL", "error")
        self._tfs_default_model_name = os.environ.get("SAGEMAKER_TFS_DEFAULT_MODEL_NAME", "None")
        self._sagemaker_port_range = os.environ.get("SAGEMAKER_SAFE_PORT_RANGE", None)
        self._tfs_config_path = "/sagemaker/model-config.cfg"
        self._tfs_batching_config_path = "/sagemaker/batching-config.cfg"

        _enable_batching = os.environ.get("SAGEMAKER_TFS_ENABLE_BATCHING", "false").lower()
        _enable_multi_model_endpoint = os.environ.get("SAGEMAKER_MULTI_MODEL",
                                                      "false").lower()

        if _enable_batching not in ["true", "false"]:
            raise ValueError("SAGEMAKER_TFS_ENABLE_BATCHING must be 'true' or 'false'")
        self._tfs_enable_batching = _enable_batching == "true"

        if _enable_multi_model_endpoint not in ["true", "false"]:
            raise ValueError("SAGEMAKER_MULTI_MODEL must be 'true' or 'false'")
        self._tfs_enable_multi_model_endpoint = _enable_multi_model_endpoint == "true"

        self._use_gunicorn = self._enable_python_service or self._tfs_enable_multi_model_endpoint

        if self._sagemaker_port_range is not None:
            parts = self._sagemaker_port_range.split("-")
            low = int(parts[0])
            hi = int(parts[1])
            if low + 2 > hi:
                raise ValueError("not enough ports available in SAGEMAKER_SAFE_PORT_RANGE ({})"
                                 .format(self._sagemaker_port_range))
            self._tfs_grpc_port = str(low)
            self._tfs_rest_port = str(low + 1)
        else:
            # just use the standard default ports
            self._tfs_grpc_port = "9000"
            self._tfs_rest_port = "8501"

        # set environment variable for python service
        os.environ["TFS_GRPC_PORT"] = self._tfs_grpc_port
        os.environ["TFS_REST_PORT"] = self._tfs_rest_port

    def _create_tfs_config(self):
        models = tfs_utils.find_models()
        if not models:
            raise ValueError("no SavedModel bundles found!")

        if self._tfs_default_model_name == "None":
            default_model = os.path.basename(models[0])
            if default_model:
                self._tfs_default_model_name = default_model
                log.info("using default model name: {}".format(self._tfs_default_model_name))
            else:
                log.info("no default model detected")

        # config (may) include duplicate 'config' keys, so we can't just dump a dict
        config = "model_config_list: {\n"
        for m in models:
            config += "  config: {\n"
            config += "    name: '{}',\n".format(os.path.basename(m))
            config += "    base_path: '{}',\n".format(m)
            config += "    model_platform: 'tensorflow'\n"
            config += "  }\n"
        config += "}\n"

        log.info("tensorflow serving model config: \n%s\n", config)

        with open("/sagemaker/model-config.cfg", "w") as f:
            f.write(config)

    def _setup_gunicorn(self):
        python_path_content = []
        python_path_option = ""

        if self._enable_python_service:
            lib_path_exists = os.path.exists(PYTHON_LIB_PATH)
            requirements_exists = os.path.exists(REQUIREMENTS_PATH)
            python_path_content = ["/opt/ml/model/code"]
            python_path_option = "--pythonpath "

            if lib_path_exists:
                python_path_content.append(PYTHON_LIB_PATH)

            if requirements_exists:
                if lib_path_exists:
                    log.warning("loading modules in '{}', ignoring requirements.txt"
                                .format(PYTHON_LIB_PATH))
                else:
                    log.info("installing packages from requirements.txt...")
                    pip_install_cmd = "pip3 install -r {}".format(REQUIREMENTS_PATH)
                    try:
                        subprocess.check_call(pip_install_cmd.split())
                    except subprocess.CalledProcessError:
                        log.error("failed to install required packages, exiting.")
                        self._stop()
                        raise ChildProcessError("failed to install required packages.")

        gunicorn_command = (
            "gunicorn -b unix:/tmp/gunicorn.sock -k gevent --chdir /sagemaker "
            "{}{} -e TFS_GRPC_PORT={} -e SAGEMAKER_MULTI_MODEL={} -e SAGEMAKER_SAFE_PORT_RANGE={} "
            "python_service:app").format(python_path_option, ",".join(python_path_content),
                                         self._tfs_grpc_port, self._tfs_enable_multi_model_endpoint,
                                         self._sagemaker_port_range)

        log.info("gunicorn command: {}".format(gunicorn_command))
        self._gunicorn_command = gunicorn_command

    def _create_nginx_config(self):
        template = self._read_nginx_template()
        pattern = re.compile(r"%(\w+)%")

        template_values = {
            "TFS_VERSION": self._tfs_version,
            "TFS_REST_PORT": self._tfs_rest_port,
            "TFS_DEFAULT_MODEL_NAME": self._tfs_default_model_name,
            "NGINX_HTTP_PORT": self._nginx_http_port,
            "NGINX_LOG_LEVEL": self._nginx_loglevel,
            "FORWARD_PING_REQUESTS": GUNICORN_PING if self._use_gunicorn else JS_PING,
            "FORWARD_INVOCATION_REQUESTS": GUNICORN_INVOCATIONS if self._use_gunicorn
            else JS_INVOCATIONS,
        }

        config = pattern.sub(lambda x: template_values[x.group(1)], template)
        log.info("nginx config: \n%s\n", config)

        with open("/sagemaker/nginx.conf", "w") as f:
            f.write(config)

    def _read_nginx_template(self):
        with open("/sagemaker/nginx.conf.template", "r") as f:
            template = f.read()
            if not template:
                raise ValueError("failed to read nginx.conf.template")

            return template

    def _start_tfs(self):
        self._log_version("tensorflow_model_server --version', 'tensorflow version info:")
        cmd = tfs_utils.tfs_command(
            self._tfs_grpc_port,
            self._tfs_rest_port,
            self._tfs_config_path,
            self._tfs_enable_batching,
            self._tfs_batching_config_path,
        )
        log.info("tensorflow serving command: {}".format(cmd))
        p = subprocess.Popen(cmd.split())
        log.info("started tensorflow serving (pid: %d)", p.pid)
        self._tfs = p

    def _start_gunicorn(self):
        self._log_version("gunicorn --version", "gunicorn version info:")
        env = os.environ.copy()
        env["TFS_DEFAULT_MODEL_NAME"] = self._tfs_default_model_name
        p = subprocess.Popen(self._gunicorn_command.split(), env=env)
        log.info("started gunicorn (pid: %d)", p.pid)
        self._gunicorn = p

    def _start_nginx(self):
        self._log_version("/usr/sbin/nginx -V", "nginx version info:")
        p = subprocess.Popen("/usr/sbin/nginx -c /sagemaker/nginx.conf".split())
        log.info("started nginx (pid: %d)", p.pid)
        self._nginx = p

    def _log_version(self, command, message):
        try:
            output = subprocess.check_output(
                command.split(),
                stderr=subprocess.STDOUT).decode("utf-8", "backslashreplace").strip()
            log.info("{}\n{}".format(message, output))
        except subprocess.CalledProcessError:
            log.warning("failed to run command: %s", command)

    def _stop(self, *args):  # pylint: disable=W0613
        self._state = "stopping"
        log.info("stopping services")
        try:
            os.kill(self._nginx.pid, signal.SIGQUIT)
        except OSError:
            pass
        try:
            if self._gunicorn:
                os.kill(self._gunicorn.pid, signal.SIGTERM)
        except OSError:
            pass
        try:
            os.kill(self._tfs.pid, signal.SIGTERM)
        except OSError:
            pass

        self._state = "stopped"
        log.info("stopped")

    def _wait_for_gunicorn(self):
        while True:
            if os.path.exists("/tmp/gunicorn.sock"):
                log.info("gunicorn server is ready!")
                return

    @contextmanager
    def _timeout(self, seconds):
        def _raise_timeout_error(signum, frame):
            raise TimeoutError("time out after {} seconds".format(seconds))

        try:
            signal.signal(signal.SIGALRM, _raise_timeout_error)
            signal.alarm(seconds)
            yield
        finally:
            signal.alarm(0)

    def start(self):
        log.info("starting services")
        self._state = "starting"
        signal.signal(signal.SIGTERM, self._stop)

        self._create_nginx_config()

        if self._tfs_enable_batching:
            log.info("batching is enabled")
            tfs_utils.create_batching_config(self._tfs_batching_config_path)

        if self._tfs_enable_multi_model_endpoint:
            log.info("multi-model endpoint is enabled, TFS model servers will be started later")
        else:
            tfs_utils.create_tfs_config(
                self._tfs_default_model_name,
                self._tfs_config_path
            )
            self._create_tfs_config()
            self._start_tfs()

        if self._use_gunicorn:
            self._setup_gunicorn()
            self._start_gunicorn()
            # make sure gunicorn is up
            with self._timeout(seconds=30):
                self._wait_for_gunicorn()

        self._start_nginx()
        self._state = "started"

        while True:
            pid, status = os.wait()

            if self._state != "started":
                break

            if pid == self._nginx.pid:
                log.warning("unexpected nginx exit (status: {}). restarting.".format(status))
                self._start_nginx()

            elif pid == self._tfs.pid:
                log.warning(
                    "unexpected tensorflow serving exit (status: {}). restarting.".format(status))
                self._start_tfs()

            elif self._gunicorn and pid == self._gunicorn.pid:
                log.warning("unexpected gunicorn exit (status: {}). restarting."
                            .format(status))
                self._start_gunicorn()

        self._stop()


if __name__ == "__main__":
    ServiceManager().start()
